main/lsp/requests/type_definition.cc

```diff
@@ -0,0 +1,72 @@
+#include "common/typecase.h"
+#include "core/lsp/QueryResponse.h"
+#include "main/lsp/lsp.h"
+
+using namespace std;
+
+namespace sorbet::realmain::lsp {
+
+namespace {
+vector<core::Loc> locsForType(const core::GlobalState &gs, core::TypePtr type) {
+    vector<core::Loc> result;
+    if (type->isUntyped()) {
+        return result;
+    }
+    typecase(
+        type.get(), [&](core::ClassType *t) { result.emplace_back(t->symbol.data(gs)->loc()); },
+        [&](core::AppliedType *t) { result.emplace_back(t->klass.data(gs)->loc()); },
+        [&](core::OrType *t) {
+            for (auto loc : locsForType(gs, t->left)) {
+                result.emplace_back(loc);
+            }
+            for (auto loc : locsForType(gs, t->right)) {
+                result.emplace_back(loc);
+            }
+        },
+        [&](core::AndType *t) {
+            for (auto loc : locsForType(gs, t->left)) {
+                result.emplace_back(loc);
+            }
+            for (auto loc : locsForType(gs, t->right)) {
+                result.emplace_back(loc);
+            }
+        });
+    return result;
+}
+} // namespace
+
+LSPResult LSPLoop::handleTextDocumentTypeDefinition(unique_ptr<core::GlobalState> gs, const MessageId &id,
+                                                    const TextDocumentPositionParams &params) const {
+    auto response = make_unique<ResponseMessage>("2.0", id, LSPMethod::TextDocumentTypeDefinition);
+    prodCategoryCounterInc("lsp.messages.processed", "textDocument.typeDefinition");
+    auto result = setupLSPQueryByLoc(move(gs), params.textDocument->uri, *params.position,
+                                     LSPMethod::TextDocumentTypeDefinition, false);
+    gs = move(result.gs);
+    if (result.error) {
+        // An error happened while setting up the query.
+        response->error = move(result.error);
+    } else {
+        auto &queryResponses = result.responses;
+        vector<unique_ptr<Location>> result;
+        if (!queryResponses.empty()) {
+            const bool fileIsTyped =
+                config.uri2FileRef(*gs, params.textDocument->uri).data(*gs).strictLevel >= core::StrictLevel::True;
+            auto resp = move(queryResponses[0]);
+
+            // Only support go-to-type-definition on constants in untyped files.
+            if (resp->isConstant() || (fileIsTyped && (resp->isIdent() || resp->isLiteral()))) {
+                for (auto loc : locsForType(*gs, resp->getRetType())) {
+                    addLocIfExists(*gs, result, loc);
+                }
+            } else if (fileIsTyped && resp->isSend()) {
+                auto sendResp = resp->isSend();
+                for (auto loc : locsForType(*gs, sendResp->dispatchResult->returnType)) {
```

Comment:
Hmm, is this the expected semantics? I'd expect go-to-type-definition on a method to go to the method, which is what typescript does. It's somewhat surprising that it jumps to the definition site of the return value.

Reply:
Oh interesting.

It seems that if you're on a send, "Go to Definition" should go to the method's definition, and "Go to Type Definition" should then go to the definition of the type that method returns. I kind of seeing it as skipping a jump: rather than having to "Go to Definition", find the sig, and then "Go to Definition" on the `returns(...)` within the sig, "Go to Type Definition" goes there in one jump.

I kind of built this feature w.r.t. what features I wished i had when using e.g. clangd, rather than looking at what other LSP clients did.